File: frontend/src/model/trace.ts
export type DisclosureLevel = "behavior" | "logic" | "function" | "dataflow" | "statement";

export type RuntimeEvidence = {
  executed: boolean;
  start_time?: number | null;
  end_time?: number | null;
  input_values: Record<string, unknown>;
  output_values: Record<string, unknown>;
  exception?: string | null;
};

export type ProgramNode = {
  id: string;
  label: string;
  level: DisclosureLevel;
  parent_id?: string | null;
  file?: string | null;
  start_line?: number | null;
  end_line?: number | null;
  children: string[];
  incoming: string[];
  outgoing: string[];
  dataflow_inputs: string[];
  dataflow_outputs: string[];
  expression?: string | null;
  runtime: RuntimeEvidence;
};

export type AgentEventType =
  | "search"
  | "open_file"
  | "symbol_lookup"
  | "inspect_function"
  | "run_test"
  | "backtrack"
  | "patch"
  | "execute";

export type AgentEvent = {
  id: string;
  timestamp: number;
  type: AgentEventType;
  target?: {
    file?: string | null;
    symbol?: string | null;
    node_id?: string | null;
  } | null;
  tool?: string | null;
  observable_input?: unknown;
  observable_output?: unknown;
};

export type ExecutionExplorationLink = {
  execution_node_id: string;
  agent_event_id: string;
  relation: "exact" | "ancestor" | "dependency" | "candidate";
  confidence?: number | null;
};

export type LineRange = {
  file: string;
  start: number;
  end: number;
};

export type ScopeContract = {
  selected_node_id: string;
  search_node_ids: string[];
  search_files: string[];
  context_node_ids: string[];
  include_runtime_values: boolean;
  include_tests: boolean;
  edit_files: string[];
  edit_line_ranges: LineRange[];
};

export type TraceBundle = {
  session_id: string;
  program_nodes: ProgramNode[];
  agent_events: AgentEvent[];
  links: ExecutionExplorationLink[];
};

File: backend/app/models/trace.py
from __future__ import annotations

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class DisclosureLevel(str, Enum):
    behavior = "behavior"
    logic = "logic"
    function = "function"
    dataflow = "dataflow"
    statement = "statement"


class RuntimeEvidence(BaseModel):
    executed: bool = False
    start_time: float | None = None
    end_time: float | None = None
    input_values: dict[str, Any] = Field(default_factory=dict)
    output_values: dict[str, Any] = Field(default_factory=dict)
    exception: str | None = None


class ProgramNode(BaseModel):
    id: str
    label: str
    level: DisclosureLevel
    parent_id: str | None = None
    file: str | None = None
    start_line: int | None = None
    end_line: int | None = None
    children: list[str] = Field(default_factory=list)
    incoming: list[str] = Field(default_factory=list)
    outgoing: list[str] = Field(default_factory=list)
    runtime: RuntimeEvidence = Field(default_factory=RuntimeEvidence)


class AgentEventType(str, Enum):
    search = "search"
    open_file = "open_file"
    symbol_lookup = "symbol_lookup"
    inspect_function = "inspect_function"
    run_test = "run_test"
    backtrack = "backtrack"
    patch = "patch"
    execute = "execute"


class AgentTarget(BaseModel):
    file: str | None = None
    symbol: str | None = None
    node_id: str | None = None


class AgentEvent(BaseModel):
    id: str
    timestamp: float
    type: AgentEventType
    target: AgentTarget | None = None
    tool: str | None = None
    observable_input: Any = None
    observable_output: Any = None


class LinkRelation(str, Enum):
    exact = "exact"
    ancestor = "ancestor"
    dependency = "dependency"
    candidate = "candidate"


class ExecutionExplorationLink(BaseModel):
    execution_node_id: str
    agent_event_id: str
    relation: LinkRelation
    confidence: float | None = Field(default=None, ge=0.0, le=1.0)


class LineRange(BaseModel):
    file: str
    start: int
    end: int


class ScopeContract(BaseModel):
    selected_node_id: str
    search_node_ids: list[str] = Field(default_factory=list)
    search_files: list[str] = Field(default_factory=list)
    context_node_ids: list[str] = Field(default_factory=list)
    include_runtime_values: bool = True
    include_tests: bool = True
    edit_files: list[str] = Field(default_factory=list)
    edit_line_ranges: list[LineRange] = Field(default_factory=list)


class TraceBundle(BaseModel):
    session_id: str
    program_nodes: list[ProgramNode] = Field(default_factory=list)
    agent_events: list[AgentEvent] = Field(default_factory=list)
    links: list[ExecutionExplorationLink] = Field(default_factory=list)
